modules/pubsub/src/main/scala/com/snowplowanalytics/snowplow/sources/pubsub/PubsubSource.scala

```diff
@@ -70,75 +68,118 @@ object PubsubSource {
+   * The toolkit for coordinating concurrent operations between the message receiver and the fs2
+   * stream. This is the interface between FS2 and non-FS2 worlds.
+   *
+   * We use pure Java (non-cats-effect) classes so we can use them in the message receiver without
+   * needing to use a cats-effect Dispatcher.
+   *
+   * @param queue
+   *   A List of messages that have been provided by the Pubsub Subscriber. The FS2 stream should
+   *   periodically drain this queue.
+   * @param semaphore
+   *   A Semaphore used to manage how many bytes are being held in memory. When the queue holds too
+   *   many bytes, the semaphore will block on acquiring more permits. This is needed because we
+   *   have turned off FlowControl in the subscriber.
+   * @param phaser
+   *   A Phaser that advances each time the queue has new messages to be consumed. The FS2 stream
+   *   can wait on the phaser instead of repeatedly pooling the queue.
+   * @param errorRef
+   *   Any error provided to us by the pubsub Subscriber. If the FS2 stream sees an error here, then
+   *   it should raise the error.
+   */
+  private case class Control(
+    queue: AtomicReference[List[SingleMessage]],
+    semaphore: Semaphore,
+    phaser: Phaser,
+    errorRef: AtomicReference[Option[Throwable]]
+  )
+
+  private object Control {
+    def build[F[_]: Sync](config: PubsubSourceConfig): F[Control] = Sync[F].delay {
+      Control(
+        new AtomicReference(Nil),
+        new Semaphore(config.bufferMaxBytes, false),
+        new Phaser(2),
+        new AtomicReference(None)
+      )
     }
   }
 
+  private def pubsubStream[F[_]: Async](config: PubsubSourceConfig): Stream[F, LowLevelEvents[Chunk[AckReplyConsumer]]] =
+    for {
+      control <- Stream.eval(Control.build(config))
+      _ <- Stream.resource(runSubscriber(config, control))
+      events <- consumeFromQueue(config, control)
+    } yield events
+
+  private def consumeFromQueue[F[_]: Sync](
+    config: PubsubSourceConfig,
+    control: Control
+  ): Stream[F, LowLevelEvents[Chunk[AckReplyConsumer]]] =
+    Stream
+      .repeatEval {
+        Sync[F].delay(control.errorRef.get).flatMap {
+          case None            => Sync[F].unit
+          case Some(throwable) => Sync[F].raiseError[Unit](throwable)
+        }
+      }
+      .evalMap { _ =>
+        // Semantically block until message receive has written at least one message
+        val waitForData = Sync[F].interruptible {
+          control.phaser.awaitAdvanceInterruptibly(control.phaser.arrive())
+        }
+        Sync[F].uncancelable { poll =>
+          poll(waitForData) *> Sync[F].delay(control.queue.getAndSet(Nil))
+        }
+      }
+      .filter(_.nonEmpty) // Would happen if phaser was terminated
+      .map { list =>
+        val events         = Chunk.iterator(list.iterator.map(_.message))
+        val acks           = Chunk.iterator(list.iterator.map(_.ackReply))
+        val earliestTstamp = list.iterator.map(_.tstamp).min
```

Comment:
I agree these three lines are unsatisfying.  It would be nice to find a neat way to traverse the list only once, whilst still building the Chunks in an efficient way, but without compromising on neatness.  I haven't found the neat way to do this yet.  I'm putting it on my personal to-do list.